### PR TITLE
Implement NOT NULL constraints for nested object columns (previously ignored)

### PR DESCRIPTION
**Pull request by matriv**
See commit messages.
Fixes: https://github.com/crate/crate/issues/5954

### CODE REVIEW

File: sql/src/main/java/io/crate/analyze/AnalyzedTableElements.java [hunk 1]
```diff
@@ -143,11 +156,26 @@ private void expandColumn(AnalyzedColumnDefinition column) {
...
+    /**
+     * Recursively add all not null constraints from child columns (object columns)
+     */
+    private void addNotNullFromChildren(AnalyzedColumnDefinition column) {
+        if (column.hasChildren()) {
+            for (AnalyzedColumnDefinition child : column.children()) {
+                if (child.hasNotNullConstraint()) {
+                    notNullColumns.add(child.ident().fqn());
+                }
+                addNotNullFromChildren(child);
```
Comment (andreidan) [inline sql/src/main/java/io/crate/analyze/AnalyzedTableElements.java]: I wouldn't use recursion for something that's easily expressed with a `while` loop (as it's a bit slower due to lack of tail recursion)

File: shared/src/main/java/io/crate/core/collections/Maps.java [hunk 2]
```diff
@@ -32,4 +32,21 @@
...
+    @SuppressWarnings("unchecked")
+    public static <T> T getNested(Map map, String key) {
+        return (T) map.get(key);
+    }
+
+    public static <T> T getNested(@Nullable Map map, String key, T defaultValue) {
+        if (map == null) {
+            return defaultValue;
+        }
+        Object o = map.get(key);
```
Comment (mfussenegger) [inline shared/src/main/java/io/crate/core/collections/Maps.java]: Any reason not to use  `map.getOrDefault()` here?

File: sql/src/main/java/io/crate/analyze/ConstraintsValidator.java [hunk 3]
```diff
@@ -22,21 +22,50 @@
...
         // Validate NOT NULL constraint
         if (value == null && !targetColumn.isNullable()) {
             throw new IllegalArgumentException(String.format(Locale.ENGLISH,
-                "Cannot insert null value for column %s", targetColumn.ident().columnIdent().fqn()));
+                "Cannot insert null value for column '%s'", targetColumn.ident().columnIdent()));
+        }
+
+        // Validate NOT NULL constraint on object nested columns
+        if (targetColumn.valueType() == DataTypes.OBJECT) {
+            Map<String, Object> valueMap = (Map<String, Object>) value;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/ConstraintsValidator.java]: could be extracted into a `validateNotNullOnChildren` method - would also remove the need for the comment

File: sql/src/main/java/io/crate/analyze/ConstraintsValidator.java [hunk 4]
```diff
@@ -22,21 +22,50 @@
...
 
 import java.util.Collection;
 import java.util.Locale;
+import java.util.Map;
 
 public final class ConstraintsValidator {
 
-    public static void validate(Object value, Reference targetColumn) {
+    @SuppressWarnings("unchecked")
+    public static void validate(Object value, Reference targetColumn, DocTableInfo tableInfo) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/ConstraintsValidator.java]: Would it make sense to change the dependency from tableInfo to just the notNullColumns?

File: sql/src/main/java/io/crate/metadata/doc/DocTableInfo.java [hunk 5]
```diff
@@ -434,4 +457,8 @@ public DynamicReference getDynamic(ColumnIdent ident, boolean forWrite) {
     public String toString() {
         return ident.fqn();
     }
+
+    public ImmutableCollection<ColumnIdent> notNullColumns() {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/doc/DocTableInfo.java]: I'd only expose this as collection, not as concrete class

File: blackbox/docs/release_notes/2.0.5.txt [hunk 6]
```diff
@@ -50,6 +50,17 @@ Changelog
 Fixes
 -----
 
+ - Implemented ``NOT NULL`` constraint validation for arbitrary level of
```
Comment (mfussenegger) [inline blackbox/docs/release_notes/2.0.5.txt]: why is this already in the release notes for 2.0.5? this fix isn't in there?

Comment (matriv) [inline blackbox/docs/release_notes/2.0.5.txt]: wanted to test with sphinx and forgot to remove, sorry...

File: shared/src/main/java/io/crate/core/collections/Maps.java (hunk 2 above)
Comment (matriv) [inline shared/src/main/java/io/crate/core/collections/Maps.java]: nope, just copied the existing code

File: sql/src/main/java/io/crate/analyze/ConstraintsValidator.java [hunk 7]
```diff
@@ -22,21 +22,52 @@
...
+    }
+
+    private static void validateNotNullOnChildren(Map<String, Object> valueMap,
+                                                  Reference targetColumn,
+                                                  Collection<ColumnIdent> notNullColumns) {
+        if (targetColumn.valueType() == DataTypes.OBJECT) {
+            for (ColumnIdent columnIdent : notNullColumns) {
+                ColumnIdent rootColumnIdent = columnIdent.getRoot();
+                // Root object column has already been validated
+                if (!rootColumnIdent.equals(columnIdent) &&
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/ConstraintsValidator.java]: would it make sense to use `ColumnIdent::isChildOf` here ?

Comment (matriv) [inline sql/src/main/java/io/crate/analyze/ConstraintsValidator.java]: yep didn't know that :-)

**Pull request merged**